### ISSUE TITLE
Creating a web interface

### ISSUE THREAD

**Issue by VictoriousFan** [simplex-chat/simplex-chat#747]
I propose creating a web interface which lets users chat by simply opening the link and clicking "Open in browser". 

The reason for this is that it would make it really easy to let new users try out the capabilities of SimpleX, and would make it easier for desktop users who don't have CLI experience to use it before a desktop app is eventually made. 

It would also  make it possible to quickly create "throwaway" secure communication channels with simply a link, something I don't know about any other way to do. The idea for this use came from me wanting to send my friend my Matrix ID, but not being able to contact them outside of an un-encrypted messenger that would leave my matrix-ID visible and analyzable in a database tied to my real identity. If I could simply send that friend a SimpleX link, without asking them to learn how to use a terminal or download an app they've never heard of before. 

Concerns: I don't know how private and secure you can make the chat when it's on a browser. This might also lead to a big increase in temporary accounts, which might bog down the servers, this could be solved by making it clear that users created in the web interface are temporary and the chat will be deleted when you close the window, which would make SimpeX a great way to open really quick secure communications channels that can be shared on any platform no matter how insecure, as long as they're the first person to use that link it's completely secure, and no user identifier is visible to any server.

**Comment by epoberezkin** [comment]
Running on the web is non-trivial - we didn't yet compile the core to the web.

Having in memory persistence for the web might be easier, as sqlite can run like that on the web. And you are right that it creates some security compromises as I think we can't do custom certification validation on the web - so it will have to be done based on domain names and certificate authorities (currently we use the approach that is similar to tor relay - including offline certificate fingerprints into server addresses.

**Comment by grinapo** [comment]
Just as a sidenote (and hopefully a useful feedback): noticing that neither native linux nor web client is on the roadmap was an immediate blocker for me for simplex.

**Comment by greensheeps** [comment]
Cpnside 

> Running on the web is non-trivial - we didn't yet compile the core to the web.
> 
> Having in memory persistence for the web might be easier, as sqlite can run like that on the web. And you are right that it creates some security compromises as I think we can't do custom certification validation on the web - so it will have to be done based on domain names and certificate authorities (currently we use the approach that is similar to tor relay - including offline certificate fingerprints into server addresses.

https://chromestatus.com/feature/5146307550248960

**Comment by p134c0d3** [comment]
Is there any update on this by chance? Since I am allowed to open things in the browser during work, it would make communicating with people I talk to constantly much easier.

**Comment by towpi13** [comment]
is this still not implemented yet? new user here this would be really helpful. it's hard for someone to download an app instead of just visiting a site

**Comment by biemster** [comment]
> is this still not implemented yet? new user here this would be really helpful. it's hard for someone to download an app instead of just visiting a site

I'm about to do this myself when time allows (not expect it anytime soon though..)

**Comment by ed-asriyan** [comment]
Since (1) there is no a single backend server and (2) the fact that JS in brower cannot open tcp connections means it would also require SMP/XFTP servers to support transport protocol over https/ws. it could raise a security topic

it's interesting though. opensource web client would allow the community to run not only community hosted servers, but community hosted clients (websites actually). 

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/3e12615b-f6f3-4bca-8943-567556c1b1cb" />

**Comment by biemster** [comment]
> Since (1) there is no a single backend server and (2) the fact that JS in brower cannot open tcp connections means it would also require SMP/XFTP servers to support transport protocol over https/ws. it could raise a security topic

I don't think it would require SMP/XFTP, my plan was to use the [CLI WebSockets API](https://github.com/simplex-chat/simplex-chat/blob/stable/bots/README.md)

**Comment by cteets-at-bluemodus** [comment]
➕ support, this could for sure capture some Signal market share. Right now, the value proposition to switch from Signal is difficult to sell to friends/family

**Comment by ed-asriyan** [comment]
> I don't think it would require SMP/XFTP, my plan was to use the [CLI WebSockets API](https://github.com/simplex-chat/simplex-chat/blob/stable/bots/README.md)

What server would the browser connect to or send messages through?

It shouldn’t rely on a web-hosted backend, as that would go against the decentralized idea of the messenger. [CLI WebSockets API](https://github.com/simplex-chat/simplex-chat/blob/stable/bots/README.md) can’t be used either, since it requires user to download and run SimpleX locally, which doesn’t really solve the original issue.

The only theoretically feasible option I can see is for a browser to connect directly to SMP/XFTP servers, same how desktop/mobile clients do. Browsers can only open https/wss connections, not any TCP. That would require updating SMP/XFTP servers code so that they support transport protocol over https/ws. Diagram: https://github.com/simplex-chat/simplex-chat/issues/747#issuecomment-3134248113

If I’m misunderstanding something, please let me know.

**Comment by biemster** [comment]
> It shouldn’t rely on a web-hosted backend, as that would go against the decentralized idea of the messenger. [CLI WebSockets API](https://github.com/simplex-chat/simplex-chat/blob/stable/bots/README.md) can’t be used either, since it requires user to download and run SimpleX locally, which doesn’t really solve the original issue.
> 
> The only theoretically feasible option I can see is for a browser to connect directly to SMP/XFTP servers, same how desktop/mobile clients do. Browsers can only open https/wss connections, not any TCP. That would require updating SMP/XFTP servers code so that they support transport protocol over https/ws. Diagram: [#747 (comment)](https://github.com/simplex-chat/simplex-chat/issues/747#issuecomment-3134248113)
> 
> If I’m misunderstanding something, please let me know.

Yes you are right, a true full web client would require SMP/XFTP, especially if you don't want to run the CLI on the backend. My use-case is different though, I just want a web interface that I can host that is controlling an already logged-in CLI, and able to message or webrtc call a couple predefined users (just me actually).

**Comment by CrypticCommit** [comment]
I've been following this with a smile and wanted to share a thought on the architecture. For many of us, a full-blown web client would be a huge step forward, provided it doesn't compromise security.

With the GHC WebAssembly backend being mature for some time now, it should be feasible to compile the core logic to Wasm. This would allow us to keep the entire E2EE process client-side. Since browsers can't open raw TCP sockets, we could use a simple WebSocket-to-TCP bridge as a relay. The bridge would only handle the transport layer and wouldn't need access to any keys — the browser would handle the actual encryption and session management locally.

This feels like the most realistic path to a web interface without needing a full rewrite. Or am I mistaken?

**Comment by ThornedRose** [comment]
Getting family to even _try_ Signal was an uphill battle. If I could tell them that they don't need a login, can just use --insert SimpleX link here-- on any browser they want to chat with me, that would make converting them to using an actual app so much easier. For example, it's been simple to get them to use Jitsi for video calls because all we do is send them a link to open in whatever browser they're already using. Easy. Getting them to install an app from the store where we have to spell the name, confirm the look of the icon, double confirm they're actually installing the correct app (and not some malware or scam app) is hard enough without having to get them to use one extra app they're not familiar with and have to explain the use of is annoyingly difficult. 

So, adding here that I also think this would be a really useful feature as a gateway for people using the app and getting more users. :) 
Sorry, I would love to be able to give practical solutions on implementing that as it's outside my skill set.

**Comment by biemster** [comment]
> If I could tell them that they don't need a login, can just use --insert SimpleX link here-- on any browser they want to chat with me, that would make converting them to using an actual app so much easier.

This is exactly my use case as well. Since I doubt this will ever happen in Simplex, I actually started making my own app.
Mobile app to send/receive messages and get phone calls, web client for family or anyone else that might need to contact me. No login or install or hoops to jump through, just a link for them. Everything local, only 3rd party dependency is a tunnel of your choosing (like cloudflare, serveo, tailscale, netbird, or a VPS)

**Comment by CrypticCommit** [comment]
I’ve continued to follow this discussion and wanted to pitch a pragmatic alternative that could give us a "Web Client" much faster, without waiting for a full native GHC WebAssembly port of the entire codebase. 

Has anyone experimented with running a **statically compiled CLI binary inside a headless WASM-emulated Linux environment running in the background**?

Instead of cross-compiling the complex Haskell core to native WASM (which is a massive undertaking due to all the crypto and SQLite dependencies), we could use an existing WASM-based virtualization layer to execute the CLI unmodified.

### Recommended Environment:
I would suggest using **[v86](https://github.com/copy/v86)** or **[WebVM](https://github.com/leaningtech/webvm)** paired with a **minimal x86_64 Alpine Linux rootfs**. Since v86 works with custom minimal Alpine images, we can reduce the initial download size to just a few megabytes for the OS layer itself. 

### The To-Do / Proof-of-Concept Steps:
To test this approach, the implementation path would look like this:

1. **Build a fully static CLI binary:** Build the `simplex-chat` CLI on an Alpine-based environment using GHCup/Cabal with the `--enable-executable-static` flag. This links all native dependencies (like `libsodium`, `gmp`, and `sqlite`) inside a single binary, making it completely independent of Ubuntu’s `glibc`.
2. **Prepare the Mini-OS Image:** Inject this static `simplex-chat` binary into a stripped-down Alpine Linux rootfs.
3. **Set up the Network Tunnel:** Since browsers restrict raw TCP sockets needed by the CLI to talk to SMP servers, configure a lightweight, stateless **WebSocket-to-TCP bridge** (like `websockify` or a basic Node.js proxy) on a backend server. 
4. **Force Proxychains inside the VM:** Use a tool like `proxychains` or `socks5` routing inside the emulated Alpine system to automatically redirect all of the CLI’s outbound TCP traffic into the WebSocket tunnel.
5. **Invisible Execution via JS:** Use the v86 JavaScript API to boot the emulator invisibly (without a canvas terminal UI) in a web worker or iframe. Pass the form input from the webpage UI directly into the VM serial port via JavaScript (`emulator.serial0_send()`).

### Why this is a powerful intermediate solution:
* **100% Client-Side Privacy:** The WebSocket-to-TCP bridge acts as a dumb transport layer. Since all SimpleX traffic is encrypted via before it even leaves the emulated CLI, the proxy server cannot inspect, read, or alter any keys or messages.
* **Perfect for Throwaway Channels:** As soon as the user closes the browser tab, the WASM RAM is completely wiped. This automatically destroys the temporary profile and database — achieving the exact "throwaway secure channel" concept mentioned in the initial issue description.

The only downside is the initial asset download size (around ~40-50MB for the static binary + mini OS), but for a zero-installation desktop web experience, this feels highly achievable right now. 

Does this architecture sound feasible to the core team, or are there underlying Haskell/network limitations that would prevent the CLI from working correctly inside an emulated environment?

**Comment by CrypticCommit** [comment]
Quick clarification: The WebSocket-to-TCP proxy would be hosted decentrally by the respective website operator running the web form (e.g., using a lightweight tool like `websockify`). This means zero infrastructure overhead or traffic cost for the SimpleX core team.

**Comment by CrypticCommit** [comment]
To follow up on this: this architecture enables a use case for anonymous website contact forms/support desks without cluttering the operator's app with single-use direct contacts.

### The Asynchronous Group-Invite Flow:
1. When the user fills out the form, the headless browser CLI automatically creates a temporary, unique **SimpleX group** (e.g., `"Support Ticket #1234"`).
2. The browser CLI sends a group invitation link directly to the operator (which gets auto-accepted) and posts the initial form message into the group.
3. The webpage UI then presents the user with a single **Group Invitation Link / QR Code** (`simplex://group-v2...`).

This solves the persistence problem perfectly: 
* The user can instantly close the tab, download the SimpleX Chat (mobile) app later, and join the group via Invitation Link whenever they want to continue the conversation.
* The operator receives the message asynchronously and can safely delete or leave the group once the customer's issue is resolved, keeping their private profile 100% clean.
